src/linear_algebra/annihilating_polynomial.lean

```diff
@@ -0,0 +1,130 @@
+/-
+Copyright (c) 2022 Justin Thomas. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Justin Thomas
+-/
+import data.polynomial
+import data.polynomial.ring_division
+import ring_theory.principal_ideal_domain
+import algebra.module.linear_map
+import field_theory.minpoly
+import linear_algebra
+import ring_theory.ideal.operations
+import ring_theory.polynomial_algebra
+
+/-!
+# Annihilating Ideal
+
+Given a commutative ring `A` and an `A`-module `M`
+(`[comm_ring A] [add_comm_group M] [module A M]`)
+Every element `u : U` defines
+an ideal `alg_hom.annihilating_ideal u ⊆ A[X]`.
+Simply put, this is the set of polynomials `p` where
+the polynomial evaluation `p(u)` is 0.
+
+## Special case where the ground ring is a field
+
+In the special case that `A` is a field, we use the notation `A = 𝕜`.
+Here `𝕜[X]` is a PID, so there is a polynomial `g ∈ alg_hom.annihilating_ideal u`
+which generates the ideal. We show that if this generator is
+chosen to be monic, then it is the minimal polynomial of `u`,
+as defined in `field_theory.minpoly`.
+
+## Special case: endomorphism algebra
+
+* `U = M →ₗ[A] M`, the endomorphism algebra of an `A`-module M.
+* `U = n × n` matrices with entries in `A`.
+-/
+
+variables {A U : Type*} [comm_semiring A] [semiring U] [algebra A U]
```

Comment:
I'd be inclined to rename `A` to `R` and `U` to `A`, since that's normally what we do for R-algebras.